src/shared_data.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+use crate::{utils, PublicKey, Result, XorName};
+use multibase::Decodable;
+use serde::{Deserialize, Serialize};
+use std::ops::Deref;
+use std::ops::Range;
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Default, Debug)]
+pub struct Key(Vec<u8>);
+
+impl Key {
+    pub fn get(&self) -> Vec<u8> {
+        self.0.to_vec()
+    }
+}
+
+impl Deref for Key {
+    type Target = Vec<u8>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<Vec<u8>> for Key {
+    fn from(vec: Vec<u8>) -> Self {
+        Key(vec)
+    }
+}
+
+// pub type Key = Vec<u8>;
+pub type Value = Vec<u8>;
+pub type KvPair = (Key, Value);
+pub type Values = Vec<Value>;
+pub type Keys = Vec<Key>;
+
+/// Marker for sentried data.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Debug)]
+pub struct Sentried;
+
+/// Marker for non-sentried data.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Debug)]
+pub struct NonSentried;
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Debug)]
+pub enum User {
+    Anyone,
+    Specific(PublicKey),
+}
+
+#[derive(Copy, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub enum Version {
+    FromStart(u64), // Absolute index
+    FromEnd(u64),   // Relative Version - start counting from the end
+}
+
+impl From<u64> for Version {
+    fn from(version: u64) -> Self {
+        Version::FromStart(version)
+    }
+}
+
+// Set of data, owners, permissions versions.
+#[derive(Copy, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub struct ExpectedVersions {
```

Comment:
It's per this design change: https://github.com/maidsafe/safe-nd/pull/126 (contains rationale)
... where recently (Sunday December 8th, in #safe-core) I discussed with @lionel1704 to change that to `Versions` instead of `Indices`:

> @lionel Regarding, the expected indices PR:
> 
> After having worked with the data types quite intensely now for a while, I've learned more about the usage of version and index at various places.
> I can see that it is used interchangeably for the same thing, where version is the concept which we IMO directly associate with the functionality of concurrency control and data perpetuity, and index is the implementation detail which reveals that we're working against a vector of sorts.
> I am quite strongly suggesting that we go consistent here and use only one of these all over, and also that it be version that we use (which was the case previously with MutableData), and keep the implementation detail from leaking through the layers.
> 
> This would in practice mean that the Index data struct is simply renamed to Version. And henceforth we talk about current version and expected version.
> I am absolutely positive that this corresponds better to the domain which it is used in (concurrency control & data perpetuity).
> Internally in safe-nd we can still map it to an Index type if we want to have a gradual transition to a language which best corresponds to that level of abstraction - i.e. when we're in the implementation details of the data types, we do know that we're dealing with indices.

where he replied

> You're absolutely right Edward. Index and version are used interchangeably. They can be renamed to version alone. Makes a lot of sense.


Please elaborate on what change you'd like to see and why.

Reply:
Ah, when I say `not much generic`, I just mean the name is `Expected`Version, meanwhile the content it holds is actually just version info, and could be used for not only `expected` purpose. (could be current, or last, ...)

Reply:
Ah OK I see. Mm that's true. I think before it was often even just an `u64`, so at least it's a bit more readable now. But I guess static type checking on the actual purpose could be implemented as well.
Do you propose an `ExpectedVersion` type, that can only be used for that purpose?

Reply:
> Do you propose an ExpectedVersion type, that can only be used for that purpose?

Yes, I'd prefer to do so.